kong/db/dao/certificates.lua

```diff
@@ -99,8 +108,26 @@ function _Certificates:update(cert_pk, cert, options)
     end
   end
 
+  local snis
+  snis, err, err_t = self.db.snis:list_for_certificate(cert_pk)
+  if not snis then
+    return nil, err, err_t
+  end
+
   -- update certificate if necessary
   if cert.key or cert.cert then
+    if cert.key == null then
+      -- user tries to remove private key from cert
+      if (name_list and #name_list ~= 0) or #snis ~= 0 then
```

Comment:
Should we distinguish those two cases?

```lua
if #name_list > 0 then
if #snis > 0 then
```

since the below error message specifies "still associated with this Certificate", but the first case means that we _want_ to associate an SNI to this Certificate, which is an important nuance for a reader of this error. 

Since `check_list_is_new()` will return `true` if all given SNIs are already associated with this Certificate, we can update the logic like such to still ensure that we differentiate both cases:
```lua
if #snis > 0 then
  -- "SNIs are still associated to this Certificate"
end

if #name_list > 0 then
  -- "cannot associate SNIs to this Certificate"
end
```

Reply:
I looked at this again, it seems that this should really be:

`if (name_list and #name_list ~= 0) and #snis ~= 0 then` instead of `or`.

It is possible for the cert to have both SNI and `key` associated with it, then user tries to update and remove both `key` and `sni` at the same time, this current logic will prevent it from happening.

Here is a test case I just added to show this, it currently fails:

```lua
it("#only removing private key and SNIs at the same time from certificate should succeed", function()
        local res = client:patch("/certificates/" .. cert_foo.id, {
          body    = {
            key = cjson.null,
            snis = cjson.null,
          },
          headers = { ["Content-Type"] = "application/json" },
        })
        local body = assert.res_status(200, res)
```

How about:

```lua
if #snis > 0 and not name_list then
  -- "SNIs are still associated to this Certificate"
end

if #name_list > 0 then
  -- "cannot associate SNIs to this Certificate"
end
```

Reply:
This commit contains the change and new tests to cover the case:

https://github.com/Kong/kong/pull/4599/commits/ec97799a4b8411331a2f0411654d249cb3d24af4

@hishamhm @thibaultcha mind taking another look? Thanks!